feat: Add JavaScript SDK migration docs to Sentry docs site (#9796)
getsentry/sentry-docs: docs/platforms/javascript/common/migration/v7-to-v8/v8-new-performance-api.mdx

---

> title: New Performance Monitoring APIs

Author: Eventually this should live under `platforms/javascript/guides/node/performance/instrumentation/custom-instrumentation/`, but this is a good spot to throw all of this in until we finally release v8.

sidebar_order: 8997
description: "Learn about new Performance Monitoring APIs in Sentry JavaScript SDK 8.x"
---

The SDK `8.x` release introduces new APIs for performance monitoring. These APIs are designed to provide more control over how performance data is collected and reported to Sentry.

> These APIs are also available in `7.x` so you can incrementally moveover to the new APIs before upgrading to `8.x`.

Reviewer: ```suggestion
These APIs are also available in `7.x` so you can incrementally move over to the new APIs before upgrading to `8.x`.
```

Previously, there were two key APIs for adding manual performance instrumentation to your applications:

- `startTransaction()`
- `span.startChild()`

> This showed the underlying data model that Sentry was originally based on, which is that there is a root Transaction which can have a nested tree of Spans.

Reviewer: ```suggestion
Those APIs were based on the original data model of Sentry. This model was based on a root Transaction that could contain a nested tree of Spans.
```

> In the new model, transactions are conceptually gone. Instead, you will always operate on spans, no matter where in the tree you are. Note that in the background, spans may still be grouped into a transaction for the Sentry UI. However, this happens transparently, and from an SDK perspective, all you have to think about are spans.

Reviewer: In the new model, transactions are conceptually gone. Now, operations are always performed on spans, regardless of their position in the tree. Note that spans may still be grouped into a transaction for the Sentry UI. However, this happens in the background and, from the SDK's point of view, all you need to be concerned about are spans.

Author: ```suggestion
In the new model, transactions are conceptually gone. Now, operations are always performed on spans, regardless of their position in the tree. Note that spans may still be grouped into a transaction for the Sentry UI. However, this happens in the background and, from the SDK's point of view, all you need to be concerned about are spans.
```

> Instead of manually starting & ending transactions and spans, the new model does not differentiate between these two. Instead, you always use the same APIs to start a new span, and it will automatically either create a new Root Span (which is just a regular span, only that it has no parent, and is thus conceptually roughly similar to a transaction) or a Child Span for whatever is the currently active span.

Reviewer: ```suggestion
The new model does not differentiate between a transaction or a span when manually starting or ending them. Instead, you always use the same APIs to start a new span. This will automatically create either a new Root Span (a regular span without a parent, conceptually similar to a transaction) or a Child Span, depending on what is the currently active span.
```

There are three key APIs available to start spans:

- `startSpan()`
- `startSpanManual()`
- `startInactiveSpan()`

All three span APIs take `StartSpanOptions` as a first argument, which has the following shape:

> ```ts

Reviewer 2: l/nit (feel free to ignore): hmm I could have sworn this worked better before but `ts`/`js` renders to 

![image](https://github.com/getsentry/sentry-docs/assets/8420481/6e2760a8-780f-4c94-904b-e2f52871fe3f)

Should we change to "JavaScript"/"TypeScript" anyway for the meantime?

interface StartSpanOptions {
  // The only required field - the name of the span
  name: string;
  attributes?: SpanAttributes;
  op?: string;
  scope?: Scope;
  forceTransaction?: boolean;
}
```

### `startSpan()`

> This is the most common API that should be used in most circumstances. It will start a new span, make it the active span

Reviewer: ```suggestion
This is the most common API that should be used in most circumstances. It will start a new span, making it the active span
```

> for the duration of a given callback, and automatically end it when the callback ends. You can use it like this:

Reviewer: ```suggestion
for the duration of a given callback, and automatically ends it when the callback ends. You can use it like this:
```

```js
Sentry.startSpan(
  {
    name: "my-span",
    attributes: {
      attr1: "my-attribute",
      attr2: 123,
    },
  },
  (span) => {
    // do something that you want to measure
    // once this is done, the span is automatically ended
  }
);
```

You can also pass an async function:

```js
Sentry.startSpan(
  {
    name: "my-span",
    attributes: {},
  },
  async (span) => {
    // do something that you want to measure
    await waitOnSomething();
    // once this is done, the span is automatically ended
  }
);
```

Since `startSpan()` will make the created span the active span, any automatic or manual instrumentation that creates
spans inside of the callback will attach new spans as children of the span we just started.

Note that if an error is thrown inside of the callback, the span status will automatically be set to be errored.

### `startSpanManual()`

> This is a variation of `startSpan()` with the only change that it does not automatically end the span when the callback
> ends, but you have to call `span.end()` yourself:

Reviewer: The start of the sentence can also be something else. I just thought it reads easier when having shorter sentences.

```suggestion
This is a variation of `startSpan()` with the only difference that it does not automatically end the span when the callback
ends. When using this, you have to call `span.end()` yourself:
```

| `span.toContext()`         | Removed                                                     |
| `span.updateWithContext()` | Removed                                                     |
| `span.getTraceContext()`   | `spanToTraceContext(span)`                                  |

In addition, a transaction has this API:

| Old name                    | Replace with                                     |
| --------------------------- | ------------------------------------------------ |
| `name`                      | `spanToJSON(span).description`                   |
| `trimEnd`                   | Removed                                          |
| `parentSampled`             | `spanIsSampled(span)` & `spanContext().isRemote` |
| `metadata`                  | Use attributes instead or set on scope           |
| `setContext()`              | Set context on scope instead                     |
| `setMeasurement()`          | `Sentry.setMeasurement()`                        |
| `setMetadata()`             | Use attributes instead or set on scope           |
| `getDynamicSamplingContext` | `getDynamicSamplingContextFromSpan(span)`        |

### Attributes vs. Data vs. Tags vs. Context

> In the old model, you had the concepts of **Data**, **Tags** and **Context** which could be used for different things.
> However, this has two main downsides: One, it is not always clear which of these should be used when. And two, not all
> of these are displayed the same way for transactions or spans.
> 
> Because of this, in the new model, there are only **Attributes** to be set on spans anymore. Broadly speaking, they map
> to what Data used to be.

Reviewer: ```suggestion
Previously, the model had the concepts of **Data**, **Tags**, and **Context**, which could be used for different things. However, this had two main downsides: Firstly, it was not always clear which concept should be used in a given situation. Secondly, these concepts were not displayed the same way for transactions or spans.  

To address these issues, the new model only has **Attributes** that can be set on spans. Broadly speaking, they map to what was formerly known as Data.
```

If you still really _need_ to set tags or context, you can do so on the scope before starting a span:

```JavaScript
Sentry.withScope((scope) => {
  scope.setTag("my-tag", "tag-value");
  Sentry.startSpan({ name: "my-span" }, (span) => {
    // do something here
    // span will have the tags from the containing scope
  });
});
```

## Other Notable Changes

In addition to generally changing the performance APIs, there are also some smaller changes that this brings with it.

### Changed `SamplingContext` for `tracesSampler()`

Currently, `tracesSampler()` can receive an arbitrary `SamplingContext` passed as argument. While this is not defined
anywhere in detail, the shape of this context will change in v8. Going forward, this will mostly receive the attributes
of the span, as well as some other relevant data of the span. Some properties we used to (sometimes) pass there, like
`req` for node-based SDKs or `location` for browser tracing, will not be passed anymore.

### No more `undefined` spans

In v7, the performance APIs `startSpan()` / `startInactiveSpan()` / `startSpanManual()` would receive an `undefined`
span if tracing was disabled or the span was not sampled.

> In v8, aligning with OpenTelemetry, these will _always_ return a span - _but_ the span may eb a Noop-Span, meaning a

Reviewer: ```suggestion
In v8, aligning with OpenTelemetry, these will _always_ return a span - _but_ the span may be a Noop-Span, meaning a
```
